app/models/cart_validations.rb

```diff
@@ -68,6 +69,18 @@ def check_overdue_reservations
     errors
   end
 
+  def check_cookie_limit
+    # checks the total number of models in the cart to prevent cookie
+    # overflow, the limit is somewhat arbitrarily set to 100 based on an
+    # estimated actual limit of >350 (see issue #880)
+    cookie_limit = 100
+    errors = []
+    if items.keys.count > cookie_limit
```

Comment:
I don't know how expensive the `keys` operation is but you can call `count` directly on a hash